product_allowed_list/models/product_list_line.py

```diff
@@ -0,0 +1,52 @@
+# Copyright 2021 Camptocamp SA
+# @author: Julien Coux <julien.coux@camptocamp.com>
+# @author: Simone Orsi <simone.orsi@camptocamp.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models
+
+
+class ProductListLine(models.Model):
+    _name = "product.allowed.list.line"
+    _description = "Product allowed list configuration Line"
+
+    product_allowed_list_id = fields.Many2one(
+        string="Product list configuration",
+        comodel_name="product.allowed.list",
+        required=True,
+    )
+    product_template_id = fields.Many2one(
+        comodel_name="product.template",
+        domain=[("sale_ok", "=", True)],
+        required=True,
+    )
+    product_id = fields.Many2one(
+        comodel_name="product.product",
+        domain="[('product_tmpl_id', '=', product_template_id)]",
+    )
+    display_name = fields.Char(compute="_compute_display_name")
+
+    def find_for_product(self, prod, config=None):
+        domain = [
+            "|",
+            ("product_id", "=", prod.id),
+            "&",
+            ("product_id", "=", False),
+            ("product_template_id", "=", prod.product_tmpl_id.id),
```

Comment:
This domain evaluates to something like:

```(product_id == prod.id) or (product_id == False and product_template_id == prod.product_tmpl_id.id)```

As we can (or at least should [^1]) trust the template will always correspond to the product_id, if present. IMO it should rather be:

```suggestion
            "|",
            ("product_id", "=", prod.id),
            ("product_id", "=", False),
            ("product_template_id", "=", prod.product_tmpl_id.id),
```
meaning:
```(product_id == prod.id or product_id == False) and product_template_id == prod.product_tmpl_id.id```


But for a shorter more readable version I'd go for this:

```suggestion
            ("product_id", "in", [False, prod.id]),
            ("product_template_id", "=", prod.product_tmpl_id.id),
```


[^1]: It may be a good idea to add an `_sql_constraint` for this